### PR TITLE
Fix test suite failure reporting with python 3.11

### PR DESCRIPTION
**Pull request by juj**
This PR fixes a failure, if a test fails on Windows, instead of giving the actual callstack of the failed test, it would report

```
DONE: combining results on main thread

Traceback (most recent call last):
  File "E:\code\emsdk\emscripten\em script\test\\runner.py", line 467, in <module>
    sys.exit(main(sys.argv))
             ^^^^^^^^^^^^^^
  File "E:\code\emsdk\emscripten\em script\test\\runner.py", line 456, in main
    num_failures = run_tests(options, suites)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\code\emsdk\emscripten\em script\test\\runner.py", line 332, in run_tests
    res = testRunner.run(suite)
          ^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\unittest\runner.py", line 217, in run
    test(result)
  File "C:\Python311\Lib\unittest\suite.py", line 84, in __call__
    return self.run(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\code\emsdk\emscripten\em script\test\parallel_testsuite.py", line 64, in run
    return self.combine_results(result, results)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "E:\code\emsdk\emscripten\em script\test\parallel_testsuite.py", line 86, in combine_results
    r.updateResult(result)
  File "E:\code\emsdk\emscripten\em script\test\parallel_testsuite.py", line 104, in updateResult
    self.buffered_result.updateResult(result)
  File "E:\code\emsdk\emscripten\em script\test\parallel_testsuite.py", line 172, in updateResult
    result.addFailure(self.test, self.error)
  File "C:\Python311\Lib\unittest\runner.py", line 106, in addFailure
    super(TextTestResult, self).addFailure(test, err)
  File "C:\Python311\Lib\unittest\result.py", line 17, in inner
    return method(self, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\unittest\result.py", line 122, in addFailure
    self.failures.append((test, self._exc_info_to_string(err, test)))
                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\unittest\result.py", line 177, in _exc_info_to_string
    tb_e = traceback.TracebackException(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\traceback.py", line 690, in __init__
    self.stack = StackSummary._extract_from_extended_frame_gen(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Python311\Lib\traceback.py", line 416, in _extract_from_extended_frame_gen
    for f, (lineno, end_lineno, colno, end_colno) in frame_gen:
  File "C:\Python311\Lib\traceback.py", line 353, in _walk_tb_with_full_positions
    positions = _get_code_position(tb.tb_frame.f_code, tb.tb_lasti)
                                                       ^^^^^^^^^^^
AttributeError: 'FakeTraceback' object has no attribute 'tb_lasti'
```

### CODE REVIEW

File: test/parallel_testsuite.py [hunk 1]
```diff
@@ -165,23 +165,41 @@ def updateResult(self, result):
...
+  def wrap_inside_function(data):
+    def fn():
+      return data
+    return fn
+  ex = fake_exc[2]
+  while ex is not None:
+    # .co_positions is supposed to be a function that returns an enumerable
+    # to the list of code positions. Create a function object wrapper around
+    # the data
+    ex.tb_frame.f_code.co_positions = wrap_inside_function(ex.tb_frame.f_code.positions)
```
Comment (sbc100) [inline test/parallel_testsuite.py:178]: I think you can just do `ex.tb_frame.f_code.co_positions = lambda: ex.tb_frame.f_code.positions`

**Comment by sbc100** [comment]
The code in `traceback.py` looks fairly generic (`positions = _get_code_position(tb.tb_frame.f_code, tb.tb_lasti)`... is this problem limited to windows or is this a general issue with python 3.11?

**Comment by sbc100** [comment]
Confirmed this is not a windows issue but a python3.11 issue:

```
test_hello_world (test_core.core0.test_hello_world) ... Traceback (most recent call last):
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/./test/runner.py", line 466, in <module>
    sys.exit(main(sys.argv))
             ^^^^^^^^^^^^^^
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/./test/runner.py", line 455, in main
    num_failures = run_tests(options, suites)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/./test/runner.py", line 331, in run_tests
    res = testRunner.run(suite)
          ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/unittest/runner.py", line 217, in run
    test(result)
  File "/usr/lib/python3.11/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/test/parallel_testsuite.py", line 62, in run
    return self.combine_results(result, results)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/test/parallel_testsuite.py", line 84, in combine_results
    r.updateResult(result)
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/test/parallel_testsuite.py", line 102, in updateResult
    self.buffered_result.updateResult(result)
  File "/usr/local/google/home/sbc/dev/wasm/emscripten/test/parallel_testsuite.py", line 180, in updateResult
    result.addError(self.test, self.error)
  File "/usr/lib/python3.11/unittest/runner.py", line 98, in addError
    super(TextTestResult, self).addError(test, err)
  File "/usr/lib/python3.11/unittest/result.py", line 17, in inner
    return method(self, *args, **kw)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/unittest/result.py", line 115, in addError
    self.errors.append((test, self._exc_info_to_string(err, test)))
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/unittest/result.py", line 177, in _exc_info_to_string
    tb_e = traceback.TracebackException(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/traceback.py", line 690, in __init__
    self.stack = StackSummary._extract_from_extended_frame_gen(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/traceback.py", line 416, in _extract_from_extended_frame_gen
    for f, (lineno, end_lineno, colno, end_colno) in frame_gen:
  File "/usr/lib/python3.11/traceback.py", line 353, in _walk_tb_with_full_positions
    positions = _get_code_position(tb.tb_frame.f_code, tb.tb_lasti)
                                                       ^^^^^^^^^^^
AttributeError: 'FakeTraceback' object has no attribute 'tb_lasti'
```

File: test/parallel_testsuite.py (hunk 1 above)
Comment (juj) [reply test/parallel_testsuite.py]: Thanks! Cleaner that way

**Comment by sbc100** [comment]
Can you title something like "Fix test suite failure reporting with python 3.11"?

**Comment by sbc100** [comment]
BTW, I guess we should start doing some minimal testing with the oldest supported version of python and very latest version (like we do now for node).

**Comment by juj** [comment]
> Can you title something like "Fix test suite failure reporting with python 3.11"?

check

> BTW, I guess we should start doing some minimal testing with the oldest supported version of python and very latest version (like we do now for node).

Yeah, that might be nice.. python can be brittle from version to version.

Looks like there are still cases that the mocking does not take into account, CI is showing real failures. I'll have to look into those a bit more.

**Comment by sbc100** [comment]
I ran into this today when my system python updated to python3.11 :(

Any chance you might have time to look into the remaining failure?

**Comment by juj** [comment]
Ops, yeah, I should take a look.. I recall I spent an hour or two trying to figure out how to create the suitable mocked structure for the same format, but wasn't able to get it going. Thanks for landing the fix independently, I'll close this down in the meanwhile until I can figure it out.

**Pull request closed without merging**